File: clients/windows/src/TNestVPN.Core/EnrollmentClient.cs
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text.Json;
using NSec.Cryptography;

namespace TNestVPN.Core;

public sealed record EnrollmentDocument(
    int Version,
    string Type,
    string ManagementUrl,
    string Token,
    DateTimeOffset ExpiresAt,
    string DeviceName,
    string? Purpose);

public sealed class EnrollmentClient(HttpClient http)
{
    private static readonly Uri TrustedOrigin = new("https://vpn.example.com");
    private static readonly JsonSerializerOptions Json = new(JsonSerializerDefaults.Web);

    public async Task<WireGuardConfig> EnrollAsync(string document, CancellationToken cancellation)
    {
        var invite = JsonSerializer.Deserialize<EnrollmentDocument>(document, Json)
            ?? throw new FormatException("注册文件无效");
        var management = new Uri(invite.ManagementUrl);
        if (invite.Version != 1 || invite.Type != "tnest-vpn-enrollment" ||
            management.Scheme != Uri.UriSchemeHttps || management.Host != TrustedOrigin.Host ||
            management.Port != 443 || management.AbsolutePath != "/" ||
            invite.ExpiresAt <= DateTimeOffset.UtcNow || invite.Token.Length < 43)
            throw new FormatException("注册文件无效或已过期");

        var algorithm = KeyAgreementAlgorithm.X25519;
        using var privateKey = new Key(algorithm, new KeyCreationParameters {
            ExportPolicy = KeyExportPolicies.AllowPlaintextExport
        });
        var privateBytes = privateKey.Export(KeyBlobFormat.RawPrivateKey);
        var publicBytes = privateKey.PublicKey.Export(KeyBlobFormat.RawPublicKey);
        var psk = RandomNumberGenerator.GetBytes(32);
        try
        {
            using var response = await http.PostAsJsonAsync(
                new Uri(TrustedOrigin, "/api/v1/enrollments/claim"),
                new {
                    token = invite.Token,
                    public_key = Convert.ToBase64String(publicBytes),
                    preshared_key = Convert.ToBase64String(psk)
                }, Json, cancellation);
            response.EnsureSuccessStatusCode();
            using var payload = await JsonDocument.ParseAsync(
                await response.Content.ReadAsStreamAsync(cancellation),
                cancellationToken: cancellation);
            var root = payload.RootElement;
            var config = root.GetProperty("configuration");
            var peer = config.GetProperty("peer");
            var addresses = string.Join(", ", config.GetProperty("address")
                .EnumerateArray().Select(x => x.GetString()));
            var dns = string.Join(", ", config.GetProperty("dns")
                .EnumerateArray().Select(x => x.GetString()));
            var rendered = $"""
                [Interface]
                PrivateKey = {Convert.ToBase64String(privateBytes)}
                Address = {addresses}
                DNS = {dns}
                MTU = {config.GetProperty("mtu").GetInt32()}

                [Peer]
                PublicKey = {peer.GetProperty("server_public_key").GetString()}
                PresharedKey = {Convert.ToBase64String(psk)}
                Endpoint = {peer.GetProperty("endpoint").GetString()}
                AllowedIPs = 0.0.0.0/0, ::/0
                PersistentKeepalive = 25

                """;
            return SafeWireGuardConfig.Parse(rendered);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(privateBytes);
            CryptographicOperations.ZeroMemory(psk);
        }
    }
}

File: clients/android/app/src/main/java/asia/tnestai/vpn/EnrollmentClient.kt
package asia.tnestai.vpn

import com.wireguard.crypto.Key
import com.wireguard.crypto.KeyPair
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.security.SecureRandom
import java.time.Instant

class EnrollmentClient {
    data class Result(val config: String, val deviceName: String)

    fun enroll(document: String): Result {
        val invite = JSONObject(document)
        require(invite.length() in 6..7)
        require(invite.getInt("version") == 1)
        require(invite.getString("type") == "tnest-vpn-enrollment")
        val origin = URI(invite.getString("management_url"))
        require(origin.scheme == "https" && origin.host == "vpn.example.com" &&
                origin.rawPath.orEmpty() in listOf("", "/") && origin.rawQuery == null)
        require(Instant.parse(invite.getString("expires_at")).isAfter(Instant.now()))
        val token = invite.getString("token")
        require(token.length >= 43)
        val pair = KeyPair()
        val pskBytes = ByteArray(32).also(SecureRandom()::nextBytes)
        val psk = Key.fromBytes(pskBytes)
        pskBytes.fill(0)

        val request = JSONObject()
            .put("token", token)
            .put("public_key", pair.publicKey.toBase64())
            .put("preshared_key", psk.toBase64())
        val connection = URL("${origin.scheme}://${origin.host}/api/v1/enrollments/claim")
            .openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.connectTimeout = 10_000
        connection.readTimeout = 20_000
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json")
        connection.outputStream.use { it.write(request.toString().toByteArray()) }
        require(connection.responseCode == 200) { "注册失败（HTTP ${connection.responseCode}）" }
        val response = connection.inputStream.bufferedReader().use { JSONObject(it.readText()) }
        val device = response.getJSONObject("device")
        val config = response.getJSONObject("configuration")
        val peer = config.getJSONObject("peer")
        val addresses = config.getJSONArray("address").let { "${it.getString(0)}, ${it.getString(1)}" }
        val dns = config.getJSONArray("dns").let { array ->
            (0 until array.length()).joinToString(", ") { array.getString(it) }
        }
        val rendered = """
            [Interface]
            PrivateKey = ${pair.privateKey.toBase64()}
            Address = $addresses
            DNS = $dns
            MTU = ${config.getInt("mtu")}

            [Peer]
            PublicKey = ${peer.getString("server_public_key")}
            PresharedKey = ${psk.toBase64()}
            Endpoint = ${peer.getString("endpoint")}
            AllowedIPs = 0.0.0.0/0, ::/0
            PersistentKeepalive = 25
        """.trimIndent() + "\n"
        SafeConfig.parse(rendered)
        return Result(rendered, device.getString("name"))
    }
}
